Plugins/Skinning/src/PositionBasedSkinning/PBS.cpp

```diff
@@ -0,0 +1,521 @@
+    }
+}
+
+/* ----------------------
+  Add Constraints
+ ----------------------- */
+
+void PBS::addConstraint_stretch ()
+{
+    unsigned int endPoint1, endPoint2;
+    TetModel *tetrahedra = m_model.getTetModel();
+    TetModel::Edges set_edges = tetrahedra->getEdges();
+
+    // Parallelisme ?
+    for (auto it_edges = set_edges.cbegin(); it_edges != set_edges.cend(); ++it_edges)
+    {
+       endPoint1 = it_edges->m_vert[0];
+       endPoint2 = it_edges->m_vert[1];
+       m_model.addDistanceConstraint(endPoint1, endPoint2);
+    }
+}
+
+void PBS::addConstraint_volume ()
+{
+    TetModel *tetrahedra = m_model.getTetModel();
+    TetModel::Tets tets = tetrahedra->getTets();
+
+    for (auto it_tet = tets.cbegin(); it_tet != tets.cend(); it_tet += 4)
+    {
+        m_model.addVolumeConstraint(*it_tet, *(it_tet + 1), *(it_tet + 2), *(it_tet + 3));
+    }
+}
+
+
+//-----------------------------------------------------------------------------------------------------------
+
+void PBS::compute ( const Animation::Skeleton *skeleton, const Animation::Pose &pose,
+                     Ra::Core::Vector3Array &outMesh)
+{
+    if (m_volumetricMeshValid)
+    {
+        ParticleData &pd = m_model.getParticles();
+         linearBlendSkinning(pose, pd);
+         step(skeleton, pd);
+         updateVertices(outMesh);
+
+        /*for (auto i = 0; i < pd.size(); ++i)
+        {
+                 RA_DISPLAY_POINT( pd.getPosition(i), Ra::Core::Colors::Red(), 0.1f );
+        }*/
+    }
+    else
+    {
+
+        QMessageBox messageBox;
+        messageBox.critical(0,"Tetrahedral Mesh"," Tetrahedral mesh must be loaded first!");
+        messageBox.setFixedSize(500,200);
+    }
+
+}
+
+void PBS::linearBlendSkinning(const Animation::Pose &pose, ParticleData &pd)
+{
+    pd.resetCurrentPositions();
+
+    for( int k = 0; k < m_weightParticles.outerSize(); ++k )
+    {
+        const int nonZero = m_weightParticles.col( k ).nonZeros();
+        Animation::WeightMatrix::InnerIterator it0( m_weightParticles, k );
+        #pragma omp parallel for
+        for( int nz = 0; nz < nonZero; ++nz )
+        {
+            Animation::WeightMatrix::InnerIterator it = it0 + Eigen::Index(nz);
+            const uint   i = it.row();
+            const uint   j = it.col();
+            const Scalar w = it.value();
+            pd.getPosition(i) += w * ( pose[j] * pd.getPosition0(i) );
+        }
+    }
+}
+
+void PBS::clearAccelerations(ParticleData &pd)
+{
+    const unsigned int num_particles = pd.size();
+
+    for (auto i = 0; i < num_particles; i++)
+    {
+        if (pd.getMass(i) != 0.0)
+        {
+            Vector3 &a = pd.getAcceleration(i);
+            a = m_gravity;
+        }
+    }
+}
+
+void PBS::positionConstraintProjection()
+{
+    unsigned int iter = 0;
+
+    SimulationModel::ConstraintVector &constraints = m_model.getConstraints();
+    SimulationModel::ConstraintGroupVector &groups = m_model.getConstraintGroups();
```

Comment:
groups should be const right?

Reply:
correct. It can be considered as a bit of code optimization if we are passing groups, but i made it const.